### PR TITLE
[fallback]: add fallback logic

### PR DESCRIPTION
**Pull request by dfkadyr**

### CODE REVIEW

**Review by mike-diamond — APPROVED** [review]

**Review by mike-diamond — DISMISSED** [review]

File: src/modules/gql-module/graphqlFetch.ts [hunk 1]
```diff
@@ -2,18 +2,39 @@ import AbortRequest from './abortRequest'
...
 
 
+type ExtendedFetchInput<Data, Variables, ModifiedData> = FetchInput<Data, Variables, ModifiedData> & {
+  retryCount?: number
+}
+
+const sessionErrorUrls = 'errorUrls'
+
+const saveErrorUrlToSessionStorage = (url: string) => {
+  const errorUrls = JSON.parse(sessionStorage.getItem(sessionErrorUrls) || '[]')
```
[the remark this answers is not in the document]
Comment (dfkadyr) [reply src/modules/gql-module/graphqlFetch.ts]: https://github.com/stakewise/v3-sdk/pull/93/commits/b579a2c52b5429c1789b824dce33ea77e170ef33

File: src/modules/gql-module/graphqlFetch.ts [hunk 2]
```diff
@@ -2,18 +2,39 @@ import AbortRequest from './abortRequest'
...
   const operationName = query
     .replace(/^(query|mutation)\s/, '')
     .replace(/[({].*/, '')
     .trim()
 
+  const isUrlArrayValid = Array.isArray(url) && url.length > 0
+  const primaryUrl = isUrlArrayValid ? url[0] : url
+  const backupUrl = isUrlArrayValid && url.length > 1 ? url[1] : null
+
+  const errorUrls = JSON.parse(sessionStorage.getItem(sessionErrorUrls) || '[]')
```
[the remark this answers is not in the document]
Comment (dfkadyr) [reply src/modules/gql-module/graphqlFetch.ts]: https://github.com/stakewise/v3-sdk/pull/93/commits/b579a2c52b5429c1789b824dce33ea77e170ef33

File: src/modules/gql-module/graphqlFetch.ts [hunk 3]
```diff
@@ -2,18 +2,39 @@ import AbortRequest from './abortRequest'
...
+  const validUrl = errorUrls.includes(primaryUrl) ? backupUrl : primaryUrl
   const opName = operationName ? `?opName=${operationName}` : ''
-  const requestUrl = `${url}${opName}`
+  const requestUrl = `${validUrl}${opName}`
```
[the remark this answers is not in the document]
Comment (dfkadyr) [reply src/modules/gql-module/graphqlFetch.ts:54]: https://github.com/stakewise/v3-sdk/pull/93/commits/b579a2c52b5429c1789b824dce33ea77e170ef33

File: src/modules/gql-module/abortRequest.ts [hunk 4]
```diff
@@ -86,7 +103,7 @@ class AbortRequest<Data, ModifiedData> {
   }
 
   finally(callback: EmptyCallback) {
-    return this.promise.finally(callback)
+    return this.promise.catch(callback)
```
[the remark this answers is not in the document]
Comment (dfkadyr) [reply src/modules/gql-module/abortRequest.ts]: https://github.com/stakewise/v3-sdk/pull/93/commits/73e103bcaeece41109ed3cac00f2d0ad7b417a2b

File: src/modules/gql-module/graphqlFetch.ts [hunk 5]
```diff
@@ -28,6 +68,19 @@ const graphqlFetch = <Data, Variables, ModifiedData>(
     onSuccess: (data: Data) => typeof modifyResult === 'function'
       ? modifyResult(data)
       : data as unknown as ModifiedData,
+    onError: (error: Error | any) => {
```
[the remark this answers is not in the document]
Comment (dfkadyr) [reply src/modules/gql-module/graphqlFetch.ts:39]: https://github.com/stakewise/v3-sdk/pull/93/commits/3128208cf9031b3313de69f2ae88fd70ea67ab4f

File: src/modules/gql-module/graphqlFetch.ts [hunk 6]
```diff
@@ -2,18 +2,58 @@ import AbortRequest from './abortRequest'
...
 
+  const isUrlArrayValid = Array.isArray(url) && url.length > 0
+  const primaryUrl: string = isUrlArrayValid ? url[0] : url
+  const backupUrl: string | null = isUrlArrayValid && url.length > 1 ? url[1] : null
+
+  const currentErrorUrl = sessionStorage.getItem(sessionErrorUrl)
+  const validUrl = (currentErrorUrl === primaryUrl && backupUrl) ? backupUrl : primaryUrl
   const opName = operationName ? `?opName=${operationName}` : ''
-  const requestUrl = `${url}${opName}`
+  const requestUrl = `${validUrl}${opName}`
```
[the remark this answers is not in the document]
Comment (dfkadyr) [reply src/modules/gql-module/graphqlFetch.ts]: https://github.com/stakewise/v3-sdk/pull/93/commits/3128208cf9031b3313de69f2ae88fd70ea67ab4f

File: src/modules/gql-module/graphqlFetch.ts [hunk 7]
```diff
@@ -2,18 +2,58 @@ import AbortRequest from './abortRequest'
...
+      sessionStorage.setItem(sessionErrorUrl, url)
+    }
+  } catch (error) {
+    console.error('Failed to save error URL to sessionStorage:', error)
+  }
+}
+
+const clearErrorUrlInterval = () => {
+  setInterval(() => {
+    try {
```
[the remark this answers is not in the document]
Comment (dfkadyr) [reply src/modules/gql-module/graphqlFetch.ts]: https://github.com/stakewise/v3-sdk/pull/93/commits/3128208cf9031b3313de69f2ae88fd70ea67ab4f

File: src/modules/gql-module/graphqlFetch.ts [hunk 8]
```diff
@@ -2,18 +2,58 @@ import AbortRequest from './abortRequest'
...
+const sessionErrorUrl = 'errorUrl'
+
+const saveErrorUrlToSessionStorage = (url: string) => {
+  try {
+    const currentErrorUrl = sessionStorage.getItem(sessionErrorUrl)
+
+    if (currentErrorUrl !== url) {
+      sessionStorage.setItem(sessionErrorUrl, url)
+    }
+  } catch (error) {
```
[the remark this answers is not in the document]
Comment (dfkadyr) [reply src/modules/gql-module/graphqlFetch.ts]: https://github.com/stakewise/v3-sdk/pull/93/commits/3128208cf9031b3313de69f2ae88fd70ea67ab4f

File: src/modules/gql-module/graphqlFetch.ts [hunk 9]
```diff
@@ -2,18 +2,58 @@ import AbortRequest from './abortRequest'
 import type { FetchInput } from './types'
 
 
+type ExtendedFetchInput<Data, Variables, ModifiedData> = FetchInput<Data, Variables, ModifiedData> & {
+  retryCount?: number
+}
+
+const sessionErrorUrl = 'errorUrl'
```
[the remark this answers is not in the document]
Comment (dfkadyr) [reply src/modules/gql-module/graphqlFetch.ts]: https://github.com/stakewise/v3-sdk/pull/93/commits/3128208cf9031b3313de69f2ae88fd70ea67ab4f

File: src/modules/gql-module/graphqlFetch.ts [hunk 10]
```diff
@@ -2,18 +2,58 @@ import AbortRequest from './abortRequest'
...
+  const primaryUrl: string = isUrlArrayValid ? url[0] : url
+  const backupUrl: string | null = isUrlArrayValid && url.length > 1 ? url[1] : null
+
+  const currentErrorUrl = sessionStorage.getItem(sessionErrorUrl)
+  const validUrl = (currentErrorUrl === primaryUrl && backupUrl) ? backupUrl : primaryUrl
   const opName = operationName ? `?opName=${operationName}` : ''
-  const requestUrl = `${url}${opName}`
+  const requestUrl = `${validUrl}${opName}`
+
+  clearErrorUrlInterval()
```
[the remark this answers is not in the document]
Comment (dfkadyr) [reply src/modules/gql-module/graphqlFetch.ts]: https://github.com/stakewise/v3-sdk/pull/93/commits/db43920d58636d788bde577d50ee3ac43b8a123c

**Pull request merged**